File: String/3_LongestSubstringWithoutRepeatingCharacters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n = len(s)
        if n < 2:
            return n
        i, j = 0, 1
        hashmap = {s[0]: 0}
        maxlen = 1
        while i < j and j < n:
            if s[j] in hashmap and hashmap[s[j]] >= i:
                i = hashmap[s[j]] + 1
            hashmap[s[j]] = j
            j += 1
            maxlen = max(maxlen, j - i)
        return maxlen

File: Java Solutions/3_LongestSubstringWithoutRepeatingCharacters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int l = 0, res = 0;
        Map<Character, Integer> pos = new HashMap<>();
        for (int r = 0; r < s.length(); r++) {
            char ch = s.charAt(r);
            if (pos.containsKey(ch) && pos.get(ch) >= l) {
                l = pos.get(ch) + 1;
            }
            pos.put(ch, r);
            res = Math.max(res, r - l + 1);
        }
        return res;
    }
}